### PR TITLE
Fix Issue 839 [ Miscalculation in gestational age ]

### PR DESCRIPTION
**Pull request by junaidwarsivd**
merged fix-develop-error branch changes 
fixed issue# 839 miscalculation in gestational age

### CODE REVIEW

**Comment by SebaMutuku** [comment]
Resolve the build errors

File: build.gradle [hunk 1]
```diff
@@ -7,10 +7,11 @@ buildscript {
         jcenter()
         maven { url 'https://maven.fabric.io/public' }
         maven { url 'https://oss.sonatype.org/content/repositories/snapshots/' }
+        maven { url 'https://plugins.gradle.org/m2/' }
     }
     dependencies {
         classpath 'com.android.tools.build:gradle:4.0.0'
-        classpath 'org.kt3k.gradle.plugin:coveralls-gradle-plugin:2.8.3'
+        classpath 'gradle.plugin.org.kt3k.gradle.plugin:coveralls-gradle-plugin:2.10.2'
```
Comment (dubdabasoduba) [inline build.gradle:14]: Why do we need this?

File: build.gradle [hunk 2]
```diff
@@ -7,10 +7,11 @@ buildscript {
         jcenter()
         maven { url 'https://maven.fabric.io/public' }
         maven { url 'https://oss.sonatype.org/content/repositories/snapshots/' }
+        maven { url 'https://plugins.gradle.org/m2/' }
```
Comment (dubdabasoduba) [inline build.gradle:10]: Why do we need this?

File: build.gradle [hunk 3]
```diff
@@ -35,6 +34,7 @@ configure(allprojects) { project ->
             mavenCentral()
             jcenter()
             maven { url 'https://oss.sonatype.org/content/repositories/snapshots/' }
+            maven { url 'https://plugins.gradle.org/m2/'}
```
Comment (dubdabasoduba) [inline build.gradle:37]: Why do we need this?

File: opensrp-anc/build.gradle [hunk 4]
```diff
@@ -5,7 +5,7 @@ buildscript {
     }
     dependencies {
         classpath 'com.android.tools.build:gradle:4.1.3'
-        classpath 'org.kt3k.gradle.plugin:coveralls-gradle-plugin:2.8.3'
+        classpath 'gradle.plugin.org.kt3k.gradle.plugin:coveralls-gradle-plugin:2.10.2'
```
Comment (dubdabasoduba) [inline opensrp-anc/build.gradle:8]: Why do we need this?

File: opensrp-anc/build.gradle [hunk 5]
```diff
@@ -178,7 +178,7 @@ tasks.withType(Test) {
 
 dependencies {
     implementation 'androidx.appcompat:appcompat:1.3.1'
-    implementation('org.smartregister:opensrp-client-native-form:2.1.17-SNAPSHOT@aar') {
+    implementation('org.smartregister:opensrp-client-native-form:3.0.0-PREVIEW-SNAPSHOT@aar') {
```
Comment (dubdabasoduba) [inline opensrp-anc/build.gradle:181]: Why are we using a PREVIEW release here?  Why has the work not been merged to master?

File: opensrp-anc/build.gradle [hunk 6]
```diff
@@ -326,9 +326,9 @@ task javadoc(type: Javadoc) {
     classpath += configurations.compile
 }
 
-def jacocoReportPath,sourceDirs
+def jacocoReportPath, sourceDirs
 task coveralls {
-    jacocoReportPath = "${buildDir}/reports/jacoco/jacocoTestReport/merged.xml"
+    jacocoReportPath = "${buildDir}/reports/jacoco/jacocoRootReport/merged.xml"
```
Comment (dubdabasoduba) [inline opensrp-anc/build.gradle:331]: Any reasons we are changing this report folder?

File: opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java [hunk 7]
```diff
@@ -217,7 +217,7 @@ private void process(String[] mainContactForms) {
                     }
                 }
                 //Make profile always complete on second contact onwards
-                requiredFieldsMap.put(ConstantsUtils.JsonFormUtils.ANC_PROFILE_ENCOUNTER_TYPE, 0);
+               // requiredFieldsMap.put(ConstantsUtils.JsonFormUtils.ANC_PROFILE_ENCOUNTER_TYPE, 0);
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java:220]: Why are we disabling this? The workflow dictates all the required fields on Profile are filled up during the 1st contact

File: opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java [hunk 8]
```diff
@@ -623,8 +631,21 @@ private void updateDefaultValues(JSONArray stepArray, int i, JSONObject fieldObj
...
+                   if(optionsObject.has(JsonFormConstants.CONTENT_WIDGET) && secondaryValue != null)
+                   {
+                       JSONObject secondaryValueObject = new JSONObject();
+                       JSONArray  secondaryValueArray = new JSONArray();
+                       secondaryValueArray.put(secondaryValue);
+                       secondaryValueObject.put(JsonFormConstants.KEY, optionsObject.getString(JsonFormConstants.KEY));
+                       secondaryValueObject.put(JsonFormConstants.VALUES, secondaryValueArray);
+                       secondaryValueObject.put(JsonFormConstants.TYPE, optionsObject.getString(JsonFormConstants.CONTENT_WIDGET));
+                       optionsObject.put(JsonFormConstants.SECONDARY_VALUE,new JSONArray().put(secondaryValueObject));
+                   }
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java:647]: refactor this to make it reusable.

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/ANCFormUtils.java [hunk 9]
```diff
@@ -128,9 +128,14 @@ private static void processRadioButtonsSpecialWidget(JSONObject widget, List<Str
         for (int i = 0; i < jsonArray.length(); i++) {
 
             JSONObject jsonObject = jsonArray.getJSONObject(i);
-
-            if (widget.has(JsonFormConstants.VALUE) && !TextUtils.isEmpty(widget.getString(JsonFormConstants.VALUE)) &&
-                    jsonObject.getString(JsonFormConstants.KEY).equals(widget.getString(JsonFormConstants.VALUE))) {
+            String widgetValue = null;
+            if(widget.has(JsonFormConstants.VALUE))
+             widgetValue = widget.getString(JsonFormConstants.VALUE);
+            if(widgetValue != null && widgetValue.startsWith("{"))
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/util/ANCFormUtils.java:134]: Use the `!StringUtils.isBlank()` instead

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/ANCFormUtils.java [hunk 10]
```diff
@@ -128,9 +128,14 @@ private static void processRadioButtonsSpecialWidget(JSONObject widget, List<Str
...
+                widgetValue =  new JSONObject(widgetValue).getString(JsonFormConstants.VALUE);
+
+            if (widgetValue!= null && !TextUtils.isEmpty(widgetValue) &&
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/util/ANCFormUtils.java:137]: Use the `!StringUtils.isBlank() instead

File: opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java [hunk 11]
```diff
@@ -134,8 +134,12 @@ protected void populateDescendants(JSONObject jsonObject) {
         try {
             JSONObject parentField = JsonFormUtils.getFieldJSONObject(getFormStep().getJSONArray(FIELDS),
                     parents.get(jsonObject.getString(KEY)));
-            String parentId = parentField.getString(VALUE);
-
+            String parentId = "";
+            if (parentField.optString(VALUE) != null && parentField.optString(VALUE).charAt(0) == '{') {
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java:138]: Use the `!StringUtils.isBlank() instead

File: opensrp-anc/src/main/res/values-pt_BR/strings.xml [hunk 12]
```diff
@@ -1,420 +0,0 @@
-<resources>
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/res/values-pt_BR/strings.xml:1]: Why have we deleted this file?

File: opensrp-anc/src/test/java/org/smartregister/anc/library/model/LoginModelTest.java [hunk 13]
```diff
@@ -43,6 +43,7 @@ public void testIsUserLoggedOutShouldReturnTrue() {
         Assert.assertTrue(model.isUserLoggedOut());
     }
 
+    @Ignore
```
Comment (dubdabasoduba) [inline opensrp-anc/src/test/java/org/smartregister/anc/library/model/LoginModelTest.java:46]: Why are you Ignoring tests?

File: reference-app/build.gradle [hunk 14]
```diff
@@ -230,7 +230,7 @@ dependencies {
     coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.1.5'
     def powerMockVersion = '2.0.7'
     implementation project(":opensrp-anc")
-    implementation('org.smartregister:opensrp-client-native-form:2.1.17-SNAPSHOT@aar') {
+    implementation('org.smartregister:opensrp-client-native-form:2.1.16-PREVIEW-SNAPSHOT@aar') {
```
Comment (dubdabasoduba) [inline reference-app/build.gradle:233]: Why are we using a Preview version here? Why has the correct version not been published?

File: reference-app/src/main/java/org/smartregister/anc/application/AncApplication.java [hunk 15]
```diff
@@ -196,14 +196,14 @@ public Context getContext() {
     @Override
     public void onTimeChanged() {
         Utils.showToast(this, this.getString(org.smartregister.anc.library.R.string.device_time_changed));
-        context.userService().getAllSharedPreferences().saveForceRemoteLogin(true, context.allSharedPreferences().fetchRegisteredANM());
+        context.userService().getAllSharedPreferences().saveForceRemoteLogin(false, context.allSharedPreferences().fetchRegisteredANM());
```
Comment (dubdabasoduba) [inline reference-app/src/main/java/org/smartregister/anc/application/AncApplication.java:199]: Why are we changing this to `false`

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

File: opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java (hunk 7 above)
Comment (junaidwarsivd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java:220]: this line was redundant and causing the data of profile to be overwritten as this line is already invoked in processRequiredStepsField method and is called under the same method where this line was present

File: build.gradle (hunk 3 above)
Comment (junaidwarsivd) [reply build.gradle:37]: this url is added for the coverall upgrade

File: build.gradle (hunk 2 above)
Comment (junaidwarsivd) [reply build.gradle:10]: this url is added for the coverall upgrade

File: build.gradle (hunk 1 above)
Comment (junaidwarsivd) [reply build.gradle:14]: this is the coveralls plugin upgrade

File: opensrp-anc/build.gradle (hunk 4 above)
Comment (junaidwarsivd) [reply opensrp-anc/build.gradle:8]: this is the coverall plugin upgrade

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/ANCFormUtils.java (hunk 9 above)
Comment (junaidwarsivd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/util/ANCFormUtils.java]: this file has been updated

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/ANCFormUtils.java (hunk 10 above)
Comment (junaidwarsivd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/util/ANCFormUtils.java]: this file has been updated

File: opensrp-anc/src/main/res/values-pt_BR/strings.xml (hunk 12 above)
Comment (junaidwarsivd) [reply opensrp-anc/src/main/res/values-pt_BR/strings.xml:1]: this file was causing build errors as there were two pt-BR string files system does not recognize pt_BR string it recognizes pt-rbr instead

File: reference-app/src/main/java/org/smartregister/anc/application/AncApplication.java (hunk 15 above)
Comment (junaidwarsivd) [reply reference-app/src/main/java/org/smartregister/anc/application/AncApplication.java]: updated the file as suggested

File: opensrp-anc/src/test/java/org/smartregister/anc/library/model/LoginModelTest.java (hunk 13 above)
Comment (junaidwarsivd) [reply opensrp-anc/src/test/java/org/smartregister/anc/library/model/LoginModelTest.java]: this change came under the conflict resolution which is corrected

File: opensrp-anc/build.gradle (hunk 5 above)
Comment (junaidwarsivd) [reply opensrp-anc/build.gradle]: Preview is used here because the latest translation changes are not merged into master yet

**Review by SebaMutuku — DISMISSED** [review]

**Comment by SebaMutuku** [comment]
@junaidwarsivd reslove the conflicts

**Pull request closed without merging**